### PR TITLE
Rasterio Development

### PR DESCRIPTION
**Pull request by Zeitsperre**
Had to remake this branch from scratch. Be careful with push -f. 

At least we now have `git-lfs` enabled.

Checklist:
* [ ] select the internal representation of a contour (ie shapely polygon)
* [ ] io function converting shapefiles and WFS, WKT into internal representation.
* [ ] select an internal representation for the DEM/land use arrays
* [x] io function reading DEM/land use and returning internal representation.
* [x] function that takes the internal representation of the polygon and returns area
* [x] function that takes the internal rep. of DEM and returns mean elevation
* [x] function that takes the land use and returns a dict or namedtuple of land fraction

### CODE REVIEW

**Comment by Zeitsperre** [comment]
@huard I think this is good for review. I think the functionalities that we want are now properly integrated.

**Comment by huard** [comment]
Starting to work on this. Will take care of resolving the conflicts.

File: raven/processes/wps_shape_area.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+        types = ['.tar', '.zip', '.7z']
+        extensions = ['.gml', '.shp', '.geojson', '.json']  # '.gpkg' requires more handling
+
+        shape_url = archive_sniffer(shape_url, working_dir=self.workdir, archive_types=types, extensions=extensions)
+
+        properties = []
+
+        try:
+            with fiona.open(shape_url, 'r', crs=from_epsg(shape_crs)) as src:
```
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply raven/processes/wps_shape_area.py]: As it's presently written, yes. I read just now that Fiona has in recent years added some functionality to automatically sniff for an ESPG code on read inspection. I could try seeing if there's a way to use the user-entered default value in the event that the file doesn't have one.

In some rare cases, the CRS could be in a non-OGC format (ESRI tends to do this) which technically works but doesn't contain a true EPSG code. I'll have to think about how to approach that one.

File: raven/processes/wps_shape_selection.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+                    else:
+                        continue
+
+                if collect_upstream:
+                    # 'PFAF_ID' can also technically be used to described the drainage network. See HydroBASINS docs.
+                    # pfaf_start, pfaf_end = str(pfaf)[0:3], str(pfaf)[3:]
+
+                    upstream = filter(lambda f: feature['properties']['NEXT_DOWN'] == basin[0], src)
+                    for up in iter(upstream):
+                        upstream_basins.append(up['properties']['HYBAS_ID'])
```
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply raven/processes/wps_shape_selection.py]: Yes. That's the idea behind this function. The reason why I haven't included more levels is because I wasn't certain that that's what we wanted and because I wanted to evade creating a potential logic bomb that could continuously loop to return 1/4 of all listed features.

I was reading the other day that if we leverage geopandas, we might be able to perform SQL-like selection and not have to constantly re-iterate through the file.

One thing I'd like to know is if the desired end result is a union of all collected features. If so, this would drastically lower the size of the geojson output and make it a bit easier to collect.

File: raven/processes/wps_shape_selection.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                src.close()
+
+        except Exception as e:
+            msg = 'Failed to extract shape from url {}: {}'.format(shape_url, e)
+            LOGGER.error(msg)
+            return response
+
+        # response['feature'] = json.dumps(properties)
+        # response['upstream_basins'] = upstream_basins
+        response.outputs['feature'].data = json.dumps(properties)
```
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply raven/processes/wps_shape_selection.py]: Right. I forgot to add that. I'll deal with this tomorrow.

File: tests/common.py [hunk 4]
```diff
@@ -46,6 +47,15 @@
...
 TESTDATA['raven-hbv-ec-ts'] = tuple((TD / 'raven-hbv-ec').glob('Salmon-River-Near-Prince-George_*.rvt'))
 
+TESTDATA['earthenv_dem_90m'] = TD / 'earthenv_dem_90m' / 'earthenv_dem90_southernQuebec.tiff'
+TESTDATA['donnees_quebec_mrc_poly'] = TD / 'donneesqc_mrc_poly' / 'donnees_quebec_mrc_polygones.gml'
+TESTDATA['statcan_econregions_2016'] = TD / 'statcan_econregions_2016' / 'lre_000b16a_f_wgs84_quebec.gml'
+TESTDATA['watershed_vector'] = TD / 'watershed_vector' / 'LSJ_LL.zip'
+TESTDATA['hydrosheds_conditioned'] = TD / 'usgs_hydrosheds_conditioned' / 'na_dem_15s_deflate.zip'
+TESTDATA['hydrobasins_10'] = TD / 'usgs_hydrobasins' / 'hybas_lake_na_lev10_v1c.zip'
+TESTDATA['hydrobasins_11'] = TD / 'usgs_hydrobasins' / 'hybas_lake_na_lev11_v1c.zip'
+TESTDATA['hydrobasins_12'] = TD / 'usgs_hydrobasins' / 'hybas_lake_na_lev12_v1c.zip'
```
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply tests/common.py]: Good point. What I'll do is clip a small catchment area in QGIS, upload the subsets, and make my tests work off of them. In terms of the production data, should I leave it on git-lfs? I could even upload the rest of the HydroBASINS data, if we'd like.

File: raven/processes/wps_shape_selection.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (huard) [reply raven/processes/wps_shape_selection.py]: Not clear to me either, but my guess is that we need to go up all the way. 
Yes, union of all features is good.

File: tests/common.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (huard) [reply tests/common.py]: Good idea. 
Yes, leave it there for now. We can say: for production do git-lfs pull. Eventually, we'll want to link to files on geoserver I think.

**Comment by Zeitsperre** [comment]
Last push didn't work (wanted to add files via `git lfs` but it committed them directly instead). Very frustrating. This will be dealt with tomorrow.

**Comment by Zeitsperre** [comment]
My issue now has to do with garbage collection. Since the files exist in previous commits, the are in the history of the branch. Still working on cleaning this up.

**Comment by huard** [comment]
rebase ?

**Comment by Zeitsperre** [comment]
Finally cleaned this up (again). git-lfs seemed to be a bit more trouble than it was worth. Abandoning it for now and continuing with the development of generic tools. In the future, we can probably write a script linking to the data housed on our servers if we want to seriously test this out.

**Comment by Zeitsperre** [comment]
@huard 

I've made a lot of changes and could use some comments if you have the time. Processes now return addresses to temprorary (but persistent, ie: garbage collection might be worth looking into) valid `geojson` or `tiff` files and I simply need to add some better tests for edge cases. I also need to remove the remaining git-lfs files but everything seems to be working fine. Are you free to chat tomorrow about next steps?

File: raven/processes/wps_hydrobasins_shape_selection.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                         data_type='string',
+                         default='(-68.724444, 50.646667)'),
+            LiteralInput('collect_upstream',
+                         'Attempt to capture both containing basin and all upstream basins from point',
+                         data_type='boolean',
+                         default='false'),
+            LiteralInput('crs', 'Coordinate Reference System of shape (EPSG) and lat/lon coordinates',
+                         data_type='integer',
+                         default=4326),
+            ComplexInput('shape', 'Vector Shape',
```
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply raven/processes/wps_hydrobasins_shape_selection.py]: No. The logic behind this process is specifically written for the structure of the HydroBASINS data set. I haven't come across anything that resembles the 'NEXT DOWN' identifier in any other products.

[the remark this answers is not in the document]
Comment (huard) [reply raven/processes/wps_hydrobasins_shape_selection.py]: So it makes no sense to leave that as a user parameter, correct ?

[the remark this answers is not in the document]
Comment (Zeitsperre) [reply raven/processes/wps_hydrobasins_shape_selection.py]: True. My motivation for allowing someone to change that inputs is because the HydroBASINS data is available for multiple levels and with/without lakes. Should I just replace the input to include those as hard-coded options?

[the remark this answers is not in the document]
Comment (huard) [reply raven/processes/wps_hydrobasins_shape_selection.py]: You mean level = {10,11,12} and lake = {true, false} ? That would be my suggestion, yes.

**Comment by Zeitsperre** [comment]
@huard 

Still a work in progress but I've dealt with most of the jarring problems and I would argue it can be safely merged. I'm aiming to spend a few hours writing tests throughout the coming week but everything seems to be working well enough to give functionality to raven. Perhaps once the tests are written, we can consider refactoring/restructuring some of the utilities and processes to make them more modular/readable and standardize their outputs.

*If you decide to merge the branch, please squash the commits. This will garbage collect the undesirable files that I introduced and later removed.* Cheers!

File: raven/processes/wps_shape_area.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply raven/processes/wps_shape_area.py]: CRS definitions are much better handled now. Resolving.

File: tests/common.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply tests/common.py]: Git-lfs has been full removed. The large dataset situation will need to be handled via docker image or configured on a needs-basis.

File: raven/processes/wps_shape_selection.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply raven/processes/wps_shape_selection.py]: This has been split into the `hydrobasins_shape_selection` function and is much better handled now.

**Comment by huard** [comment]
@Zeitsperre Are you planning other changes or may I continue the review ?

**Comment by Zeitsperre** [comment]
@huard Please feel free to continue. The only major changes I was considering was to refactor the GIS-specific utilities into their own file and leave the generic file-handling function in the `utills.py`. Functions seem to be stable and returning appropriate outputs from my tests.

**Comment by Zeitsperre** [comment]
Just to confirm, I'll be moving along to other projects while you review this PR. Do you think you could investigate the XML error involved in passing zip files instead of JSON output. I'm not sure how that's handled.

Also, I've noticed that there are a lot of warnings arising in the post-testing debug information. Are they worth looking into?
